llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp

```diff
@@ -6735,6 +6735,15 @@ MachineBasicBlock::iterator ARMBaseInstrInfo::insertOutlinedCall(
   return CallPt;
 }
 
+void ARMBaseInstrInfo::buildClearRegister(Register DstReg,
+                                          MachineBasicBlock &MBB,
+                                          MachineBasicBlock::iterator Iter,
+                                          DebugLoc &DL,
+                                          bool AllowSideEffects) const {
+  unsigned Opc = Subtarget.isThumb2() ? ARM::t2MOVi32imm : ARM::MOVi32imm;
```

Comment:
Should floating-point registers also be covered by this method?

Reply:
Also, do we care about Thumb1?

Reply:
> Should floating-point registers also be covered by this method?

Yes. I forgot about them. :-(

> Also, do we care about Thumb1?

I can change this to `isThumb()`, but I want to avoid using `tMOVi32imm`, because it modifies `CPSR` (I want to avoid side effects even if `AllowSideEffects` is `true`).

Reply:
> (I want to avoid side effects even if AllowSideEffects is true).

huh? Why?

I suspect you'll need to check:
if arm:
  ...
elif thumb2:
  ...
else: # is thumb1
  ...

Reply:
Avoiding side effects should always be desirable / safe. The only reason to allow them is if an alternative instruction (e.g. `xor` on x86) is smaller or more efficient or whatever. In the case of ARM / AArch64, the size of the instruction doesn't matter of course. However, ARM does produce code like this when exiting the function with stack protectors:

```
	cmp	r1, r0
	addeq	sp, sp, #8
	addeq	sp, sp, #1024
	popeq	{r11, pc}
```

So shoving a side-effect after the `cmp` instruction won't be good. Yes, I will have `AllowSideEffects == false` in that case, but I don't see why we would allow side-effects at all unless it's clearly a win.

After looking at the documentation, it *does* appear that the Thumb1 version of `MOV Rn, #imm` won't modify the condition flags, unless we specify it to. I'll make that change.